lib/streamlit/elements/utils.py

```diff
@@ -86,6 +86,39 @@ def check_session_state_rules(
         _shown_default_value_warning = True
 
 
+class CachedWidgetWarning(StreamlitAPIWarning):
+    def __init__(self):
+        super().__init__(
+            """
+Your script uses a widget command in a cached function
+(function decorated with `@st.cache_data` or `@st.cache_resource`).
+This code will only be called when we detect a cache "miss",
+which can lead to unexpected results.
+
+How to fix this:
+* Move all widget commands outside the cached function.
+* Or, if you know what you're doing, use `experimental_allow_widgets=True`
+in the cache decorator to enable widget replay and suppress this warning.
+"""
+        )
+
+
+def check_widget_usage() -> None:
```

Comment:
From reading this PR and the various places where the function is used, I believe a better name might be something like `assert_not_in_cached_function` or so. With that, someone who reads the code immediately an idea of what this function is doing without having to go through the function body.
What do you think?

Reply:
My initial intent here was to keep this method name more generic to prevent us from adding a lot of individual methods to all widgets to do similar types of checks. Keeping the number of methods we have to use in all widgets as low as possible can help to prevent a bit of bloat.